Log Run details and Refactor Random Actuator/Sensors

benderpremier opened the pull request:
This PR depends on #79  and contains 3 main features:

1.  Change the agent run method signature to `run(e: ENVIRONMENT): (ENVIRONMENT, RunDetail)`. The idea is to be able to keep track of the percepts and actions of the agent every time we call run. This will allow a UI to display details about the agent behavior.
2. Derive unreliable Sensors and Actuators from the reliable one.
3. Encode unreliable with `Option`

BusyByte commented on core/src/main/scala/aima/core/agent/Agent.scala:
```diff
@@ -3,26 +3,31 @@ package aima.core.agent
 
-trait Agent[Action, Percept] {
-  type AgentFunction = Percept => Action
+case class RunDetail[PERCEPT, ACTION](
```
@benderpremier could we call this AgentRunSummary?  Detail makes it seem like we added more information.  Additionally I do wonder if we should have some algebra to for Summary which also includes a date time of the PERCEPT or the ACTION so you could order them somehow.

benderpremier replied:
Yeah I am happy to rename this to `AgentRunSummary`. I am not a native speaker so naming stuff correctly is never easy :) . 

Isn't the order of the percepts and actions already guaranteed by the elements of the list being processed in order ?

BusyByte replied:
@benderpremier They are ordered the way it's currently implemented however I could see different implementations where they might not be.  Also it's always nice to know when the agent perceived something and when it decided to perform an action.

BusyByte commented:
@benderpremier Your other PR created merge conflicts, let me know when those are cleaned up and I'll merge this if you are happy with it.

benderpremier commented:
I have updated it. If you are ok with it I am happy for you to merge it 👍

BusyByte commented on core/src/test/scala/aima/core/environment/vacuum/ModelBasedReflexVacuumAgentSpec.scala:
```diff
@@ -24,9 +24,6 @@ class ModelBasedReflexVacuumAgentSpec extends Specification {
   "should assume dirty after moving to location B" in new context {
     agent.agentFunction.apply(LocationAPercept)
     agent.agentFunction.apply(CleanPercept)
-    agent.agentFunction.apply(NoPercept) must beLike {
```
looks like  you have an example without an assertion now
this seems like a valid example

BusyByte commented on core/src/test/scala/aima/core/environment/vacuum/ModelBasedReflexVacuumAgentSpec.scala:
```diff
@@ -45,9 +42,6 @@ class ModelBasedReflexVacuumAgentSpec extends Specification {
   "should assume dirty after moving to location A" in new context {
     agent.agentFunction.apply(LocationBPercept)
     agent.agentFunction.apply(CleanPercept)
-    agent.agentFunction.apply(NoPercept) must beLike {
```
same as above

BusyByte commented on core/src/main/scala/aima/core/environment/vacuum/SimpleReflexVacuumAgentProgram.scala:
```diff
@@ -13,6 +13,5 @@ class SimpleReflexVacuumAgentProgram extends SimpleReflexAgentProgram[VacuumPerc
     case LocationAPercept => RightMoveAction
     case LocationBPercept => LeftMoveAction
     case CleanPercept     => NoAction
-    case NoPercept        => NoAction
```
I'm not sure I like changing from NoPercept and NoAction to Option[Percept] and Option[Action] .
While it is common to represent optionality  with Option I find it harder to do things like fold.
Now I have to do map and fold and get or else whereas before I could just do fold easily.

BusyByte commented on core/src/main/scala/aima/core/agent/Actuator.scala:
```diff
@@ -1,19 +1,23 @@
   */
 trait Actuator[ENVIRONMENT, ACTION] {
-  def act(action: ACTION, e: ENVIRONMENT): ENVIRONMENT //could be Action => E => E
+  def act(action: ACTION, e: ENVIRONMENT): Option[ENVIRONMENT]
```
I'm not sure I like the idea of returning an option of an environment. 
I was thinking of the environment as immutable and this as like a state monad where if there is no action taken then it's basically identity otherwise it applies the function and gives you a new immutable state.

benderpremier replied:
Yeah I agree with you. I think the Actuator act method returning an Option isn't a good idea after all. No matter if the actuators work reliably or not, you'll always get an Environment 👍 I'll fix it.

benderpremier commented:
I have fixed the two UT (don't know how I missed it sorry about that) and reverted my changes for the actuator.

The main motivation I have for getting rid of the `NoPercept` is that you'll have to re-implement it for every Environment. Hence I left the optional for the time being. But if you prefer I can make a different PR for that. Let me know.

BusyByte commented:
@benderpremier merging
Could you try to have a bit more focused PRs where it only has changes for title of the PR?
The log summary really could have been separate from any refactoring.
